### DISCUSSION TITLE
Client operations and thread safety

### DISCUSSION THREAD

**Discussion by persicsb** [eclipse-milo/milo#1306] [General]
The client has two possible executor services set: 
- a simple ExecutorService
- a ScheduledExecutorService

What is the purpose of these executors? Are they only used by the internal FSMs?

These can be found in https://github.com/eclipse/milo/blob/master/opc-ua-stack/stack-client/src/main/java/org/eclipse/milo/opcua/stack/client/UaStackClientConfigBuilder.java#L39

Regarding #1300 , what is the expected thread safety regarding to the client?
Shall every client operation (like client.getSession(), ManagedSubscripotion.create() etc.) be executed on these executors,  or it can be issued on any thread?

The milo-examples client projects do not set these executors, and the example codes do not execute client operations on them, but on a different thread.

What is the policy here?
Shall client cconfiguration set the ExecutorService and ScheduledExecutorService ot let the client create them?
What about the client operations? Are they safe to be called on any thread?

**Comment by kevinherron** [comment]
Client and server both may use an `ExecutorService` or `ScheduledExecutorService` in its implementation.

You can also use them in your application if needed. Neither need to be configured. Some applications prefer to provide their own, even if just to control the thread names. 

`OpcUaClient`, `OpcUaSubscription`, `ManagedSubscription`, etc... are safe for use by multiple threads and do not require that you interact with them on one of the executor threads.

In the `dev/1.0` branch the client API has seen significant refactoring and the two competing Subscription APIs have been merged into one. This API is safe for use by multiple threads in terms of visibility, but not safe for concurrent use. I thought I had started documenting this but can't find it now...
